### DISCUSSION TITLE
Reading from USB serial GPS

### DISCUSSION THREAD

**Discussion by mbcoder** [Pi4J/pi4j#281] [Q&A] [answered]
I've got a USB based GPS device which I'd like to read from.  I've been following the documentation [here](https://pi4j.com/documentation/io-examples/serial/) to get me started.

I've proven my GPS device can be listened to by using Putty so I know its broadcasting on /dev/ttyUSB0 at 4800baud, 8 data bits, one stop bit.  Of course I'm closing putty before running my Java app!

My code so far looks like this:

```
System.out.println("Starting serial...");

            pi4j = Pi4J.newAutoContext();


            serial = pi4j.create(Serial.newConfigBuilder(pi4j)
                .baud(Baud._4800)
                .dataBits_8()
                .parity(Parity.NONE)
                .stopBits(StopBits._1)
                .flowControl(FlowControl.NONE)
                .id("gps port")
                .device("/dev/ttyUSB0")
                //.provider("pigpio-serial"). --  Note I commended this out as my serial is on USB port!  Different val needed?
                .build());

            System.out.println("serial " + serial);
            serial.open();

            System.out.println("about to create runnable");

            

            Runnable runnable = new Runnable() {
                @Override
                public void run() {
                    System.out.println("Waiting till serial port is open");
                    while (!serial.isOpen()) {
                        try {
                            Thread.sleep(1000);
                            System.out.println("retry...");
                        } catch (InterruptedException e) {
                            throw new RuntimeException(e);
                        }
                    }

                    //opened now!
                    System.out.println("serial port is open!");

                    // Start a thread to handle the incoming data from the serial port
                    SerialReader serialReader = new SerialReader(serial);
                    Thread serialReaderThread = new Thread(serialReader, "SerialReader");
                    serialReaderThread.setDaemon(true);
                    serialReaderThread.start();

                }
            };
            System.out.println("about to start runnable");
            runnable.run();
```

So when I run the code, it waits for the serial port to open, but it never return true when I call serial.isOpen()

The example in the documentation worked using a serial connection connected to the GPIO port, but I'm wondering if the API works for a USB based device?   Maybe I'm just doing something silly.  

Any thoughts would be much appreciated.

**Comment by taartspi** [comment, marked as answer]
You need that .provider line.  Without, the default '.build' operation has not created a usable device.    With that line included your code gets into the SerialReader thread and fails for a bad handle,   I do not have a /dev/ttyUSB0 in my tree.     Hopefully this lets you debug further.

**Comment by mbcoder** [comment]
@taartspi thanks for your reply.  I've had a dig into what the provider is and understand why its needed - basically in the background you are using the `pigpio` library.

When I added the provider back into the code, the `.build` didn't return and on killing the app, it bubbles up an exception like this:

```
Caused by: com.pi4j.library.pigpio.PiGpioException: PIGPIO ERROR: PI_INIT_FAILED; pigpio initialisation failed
        at com.pi4j.library.pigpio@2.3.0/com.pi4j.library.pigpio.impl.PiGpioBase.validateResult(PiGpioBase.java:265)
        at com.pi4j.library.pigpio@2.3.0/com.pi4j.library.pigpio.impl.PiGpioBase.validateResult(PiGpioBase.java:251)
        at com.pi4j.library.pigpio@2.3.0/com.pi4j.library.pigpio.impl.PiGpioNativeImpl.gpioInitialise(PiGpioNativeImpl.java:98)
        at com.pi4j.library.pigpio@2.3.0/com.pi4j.library.pigpio.PiGpio.initialize(PiGpio.java:159)
        at com.pi4j.plugin.pigpio@2.3.0/com.pi4j.plugin.pigpio.provider.serial.PiGpioSerialProviderImpl.create(PiGpioSerialProviderImpl.java:60)
```

I was looking around some other discussions on this forum and found that running my app with `sudo` makes it work and I don't get the exception.

So as a gradle app I run it as: `sudo ./gradlew run` and it all springs into life and I'm seeing my NMEA sentences.

So although I've got it working, is there a way of reading the serial port without running my java app from sudo?  

I can run Putty via my own user and that works, so I'm wondering of there is some config I can do to make this work?

Thanks again - this is much appreciated!

**Reply by taartspi** [reply]
Sudo is required by the pigpio library to access the hardware.  You can find conversations of people configuring the userID and pi to avoid this. I never researched if/how well it worked as my programs are used by me so adding sudo is  not a problem    Glad u have things moving.

**Comment by mbcoder** [comment]
@taartspi thanks again for your replies, I thought you'd be interested in the project I've applied this to which is in this [Blog](https://www.esri.com/arcgis-blog/products/sdk-java/developers/how-to-use-the-arcgis-maps-sdk-for-java-in-a-raspberry-pi-for-recording-gps-tracks/)

Having used the pi4j API which is an great project backed up by good documentation, I feel that I could make some improvements and additions myself so I might pop in a ticket with my ideas.

**Reply by taartspi** [reply]
Your care in documentation will encourage interested folks to give it a try.   When I read through your project my thought was to suggest you contact @FDelporte and this work be added to the Pi4j web but Frank beat me to it.   Take care, Tom

**Comment by FDelporte** [comment]
Wow @mbcoder that's a really nice post where you combine two of my "darlings": Pi4J and JavaFX ;-)
Shared it here:
https://foojay.social/@pi4j/110654372808381876
https://pi4j.com/documentation/io-examples/serial/#code-example

Will also include it in the next JFX-Central LinksOfTheWeek!

Looking forward to your feedback for the Pi4J!
